src/main/drivers/compass/compass_spi_ak8963.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <math.h>
+
+#include "platform.h"
+
+#include "build/debug.h"
+
+#include "common/axis.h"
+#include "common/maths.h"
+#include "common/utils.h"
+
+#include "drivers/bus_spi.h"
+#include "drivers/io.h"
+#include "drivers/sensor.h"
+#include "drivers/time.h"
+
+#include "drivers/compass/compass.h"
+#include "drivers/compass/compass_ak8963.h"
+#include "drivers/compass/compass_spi_ak8963.h"
+
+static float magGain[3] = { 1.0f, 1.0f, 1.0f };
+
+#define DISABLE_AK8963      IOHi(ak8963CsPin)
+#define ENABLE_AK8963       IOLo(ak8963CsPin)
+
+static IO_t ak8963CsPin = IO_NONE;
+
+static bool ak8963SpiWriteCommand(uint8_t reg, uint8_t data)
+{
+    uint8_t buf[32];
+    buf[0] = reg & 0x7F;
+    buf[1] = data;
+
+    ENABLE_AK8963;
+    spiTransfer(AK8963_SPI_INSTANCE, NULL, buf, 2);
+    DISABLE_AK8963;
+
+    return true;
+}
+
+static bool ak8963SpiReadCommand(uint8_t reg, uint8_t length, uint8_t *data)
+{
+    uint8_t buf[32];
+    buf[0] = reg | 0x80 | 0x40;
+
+    ENABLE_AK8963;
+    spiTransfer(AK8963_SPI_INSTANCE, buf, buf, length + 1);
+    DISABLE_AK8963;
+
+    memcpy(data, &buf[1], length);
+
+    return true;
+}
+
+static bool ak8963SpiInit()
+{
+    uint8_t calibration[3];
+    uint8_t status;
+
+    ak8963SpiWriteCommand(AK8963_MAG_REG_CNTL, CNTL_MODE_POWER_DOWN); // power down before entering fuse mode
+    delay(20);
+
+    ak8963SpiWriteCommand(AK8963_MAG_REG_CNTL, CNTL_MODE_FUSE_ROM); // Enter Fuse ROM access mode
+    delay(10);
+
+    ak8963SpiReadCommand(AK8963_MAG_REG_ASAX, sizeof(calibration), calibration); // Read the x-, y-, and z-axis calibration values
```

Comment:
@martinbudden : ak8963SpiReadCommand should take `void*`, no cast would be necessary. And in C arder of arguments is usually '*data', 'size' - this is quite confusing ...